support for cancellation of delete requests

sandeepsukhani opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with master
-->

**What this PR does**:
Adds support for cancellation of delete requests. Delete requests can be cancelled only when they are in `RECEIVED` status i.e it is not picked up for processing.

**NOTE**: To avoid a race between cancellation of delete requests and `DataPurger` picking up the delete request, I have added a Mutex. This is based on the assumption that `DataPurger` would always run as a single instance and it is always going to be the case since it kind of works as an orchestrator to manage delete requests. We will be scaling deletes in future by scaling workers(running as a service) not `DataPurger` service.
Alternative:
Currently, delete requests are processed only after they have aged for 24 hours. It is a hardcoded value and we are going to make it configurable just for running integration tests. Another option that we have is to allow delete requests to be cancelled only until they have aged for 24 hours and processing of delete requests to be deferred until they have aged for minimum 24 hours + 5 or so minutes. I have not considered that option because it would cause a problem in the implementation of integration tests.

**Checklist**
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented:
> To avoid a race between cancellation of delete requests and `DataPurger` picking up the delete request, I have added a Mutex. This is based on the assumption that `DataPurger` would always run as a single instance and it is always going to be the case since it kind of works as an orchestrator to manage delete requests. We will be scaling deletes in future by scaling workers(running as a service) not `DataPurger` service.

Do we need this mutex? What possible issues could happen if it wasn't there? It seems very strange to me to have a in-process lock to guard access to external database.

pstibrany commented:
> Do we need this mutex? What possible issues could happen if it wasn't there? It seems very strange to me to have a in-process lock to guard access to external database.

In addition to that, it seems to me that `pullDeleteRequestsToPlanDeletes` can take a while to execute, and while it is running, no cancellation request can proceed.

sandeepsukhani commented:
> > To avoid a race between cancellation of delete requests and `DataPurger` picking up the delete request, I have added a Mutex. This is based on the assumption that `DataPurger` would always run as a single instance and it is always going to be the case since it kind of works as an orchestrator to manage delete requests. We will be scaling deletes in future by scaling workers(running as a service) not `DataPurger` service.
> 
> Do we need this mutex? What possible issues could happen if it wasn't there? It seems very strange to me to have a in-process lock to guard access to external database.

There is no separate method in Cortex for doing check and update an index entry in index stores i.e there is only add method which does an update if a record already exists or adds it if not. This is because of the properties of underlying stores that we support. Considering that imagine someone requesting cancellation of delete request and meanwhile purger picking up that request for processing, the user would be under the impression that the delete request is cancelled successfully while purger picked up the request already due to a race and started processing it and updated the status as soon as we removed the request entry.

As I suggested the other option would be having a deadline for cancellation of delete request to avoid the race. What do you think?

pstibrany commented:
> Are we talking about deadline before scheduled delete time, until cancel request may arrive? If so, I think that may be safer. If cancel request arrives after deadline, it will be accepted, but may be ignored.

We can communicate this in API documentation and/or in API response.

sandeepsukhani commented on pkg/chunk/purger/request_handler.go:
```diff
@@ -128,3 +130,39 @@ func (dm *DeleteRequestHandler) GetAllDeleteRequestsHandler(w http.ResponseWrite
+
+	deleteRequest, err := dm.deleteStore.GetDeleteRequest(ctx, userID, requestID)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Sorry didn't get you. I am sending the error back in http response here and not logging it.
Am I missing something?

pstibrany replied:
I was just wondering, if we are returning "internal server error", if such error would be worth having in the log file as well. (I'd say yes)

sandeepsukhani replied:
yes, it is a good idea, added error logs for all the `StatusInternalServerError` responses.

sandeepsukhani commented:
> I think the "cancel the delete" request is best-effort. If it arrives too late, we cannot promise it will be fulfilled.
The problem here is not cancellation of the request being best-effort but saying the request is cancelled which actually is not due to a race.

pstibrany commented on pkg/chunk/purger/request_handler.go:
```diff
@@ -120,11 +123,53 @@ func (dm *DeleteRequestHandler) GetAllDeleteRequestsHandler(w http.ResponseWrite
+
+	if err := dm.deleteStore.RemoveDeleteRequest(ctx, userID, requestID, deleteRequest.CreatedAt, deleteRequest.StartTime, deleteRequest.EndTime); err != nil {
+		level.Error(util.Logger).Log("msg", "error cancelling the delete request", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
```
I'd suggest to end all error paths with `return` for consistency, and explicitly return `StatusOK` or `StatusNoContent`.

```suggestion
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	
	w.WriteHeader(http.StatusOK)
```

pstibrany commented on pkg/chunk/purger/purger.go:
```diff
@@ -335,7 +338,8 @@ func (dp *DataPurger) pullDeleteRequestsToPlanDeletes() error {
 
 	for _, deleteRequest := range deleteRequests {
-		if deleteRequest.CreatedAt.Add(24 * time.Hour).After(model.Now()) {
+		// adding an extra minute here to avoid a race between cancellation of request and picking of the request for processing
+		if deleteRequest.CreatedAt.Add(deleteRequestCancellationDeadline).Add(time.Hour).After(model.Now()) {
```
Comment says "minute" but code adds "hour".
